### ISSUE TITLE
Investigate implications of new Packer licensing

### ISSUE THREAD

**Issue by mboersma** [kubernetes-sigs/image-builder#1246]
HashiCorp recently changed the license on their open source project to BSL 1.1. This may or may not present legal issues to the use of Packer in the image-builder project.

Let's communicate with the CNCF to see if there's an official position on the new license, and determine whether there are implications for image-builder.

image-builder currently uses Packer v1.9.1 and is updating to v1.9.2, the last version under the MPL. The project will stay pinned there until we have a resolution to this issue.

See also:
- https://discuss.hashicorp.com/t/hashicorp-projects-changing-license-to-business-source-license-v1-1/57106
- https://www.hashicorp.com/blog/hashicorp-adopts-business-source-license
- https://www.hashicorp.com/license-faq

**Comment by AverageMarcus** [comment]
Some points of note:

* > Security fixes will be backported under MPL 2.0 through December 31, 2023.

  Taken from https://www.hashicorp.com/license-faq
* > If you need further clarification with respect to a particular use case, you can email [licensing@hashicorp.com](mailto:licensing@hashicorp.com). Custom licensing terms are also available to provide more clarity and enable use cases beyond the BSL limitations.

  Taken from https://www.hashicorp.com/license-faq
* It seems _some_ libraries and SDKs are still ok but doesn’t seem to include anything for Packer as far as I understand it. This would rule out switching to our own binary leveraging the Packer Go modules (e.g. resurrecting the cli project we recently got rid of)
* From what I understand, I don’t think the image-builder project itself would be considered a competitor but many of our users that run image-builder to create their images would be (e.g. Microsoft building for CAPZ). Maybe it would be possible for us as an OSS project to build and provide the images for the CAPI projects but that wouldn’t be ideal (cost, inflexible, maintenance burden).

**Comment by jsturtevant** [comment]
thread in the cncf on this topic: https://github.com/cncf/foundation/issues/617

**Comment by AverageMarcus** [comment]
Recommendations from CNCF: https://github.com/cncf/foundation/blob/main/source-available-recommendations.md

We've already pinned the version but thats a short term solution at best. I don't believe their is an alternative available to Packer that we could switch to, at least not without substantial work needing to be done.

**Comment by AverageMarcus** [comment]
As discussed in this slack thread: https://cloud-native.slack.com/archives/C0MP69YF4/p1692111612303149?thread_ts=1692108926.999039&cid=C0MP69YF4

We should submit a request for license exception from the CNCF Governing Board (not sure how to do that yet) just to be on the safe side. It's possible we won't need the exception but its best to check to be sure.

If no one else gets to it first I'll try and get the request in Thurs/Fri this week.

**Comment by BenTheElder** [comment]
> We should submit a request for license exception from the CNCF Governing Board (not sure how to do that yet) just to be on the safe side.

From https://github.com/cncf/foundation/blob/main/source-available-recommendations.md:

> In either case, while a project's maintainers may request a license exception from the CNCF Governing Board, it is extremely unlikely that one will be granted for the use of a dependency under a source available, non-open source license.

The CNCF has granted exceptions in the past for hashi libraries etc being under MPL, but that is still an "open source" license rather than "source available". I don't think they're likely to approve one here, the CNCF approved license list (no exceptions necessary) are all relatively free use license (ASLv2, MIT, ... https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md) and BUSL is pretty much the opposite.

That said, to file a request you file an issue at https://github.com/cncf/foundation/issues such as https://github.com/cncf/foundation/issues/187

**Comment by lentzi90** [comment]
Has anyone looked at potential alternatives to packer? I was not aware of any but did a quick search and found [linuxkit](https://github.com/linuxkit/linuxkit). Any ideas if it could replace packer? (Looks like they used to build OS images for Kubernetes with it in [this project](https://github.com/linuxkit/kubernetes/tree/master) but it hasn't been updated in 5 years.)

**Comment by mboersma** [comment]
See also kubernetes-sigs/cluster-api#9181.

**Comment by mboersma** [comment]
It appears that Packer 1.9.x patches will keep the MPL 2.0 license, although BUSL is already in the `main` branch. See #1269.

**Comment by cpboyd** [comment]
[The FAQ](https://www.hashicorp.com/license-faq#products-covered-by-bsl) notes:
> Security fixes will be backported under MPL 2.0 through December 31, 2023.

I'm wondering if towards the end of the year there might be more efforts to fork projects like packer similarly to [OpenTofu](https://github.com/opentofu/opentofu)

**Comment by BenTheElder** [comment]
I think if there were going to be a similar effort for projects other than terraform there would've been some indication of that by now.

**Comment by justinsb** [comment]
We did have alternative approaches in this project that did not require packer, though they were removed in https://github.com/kubernetes-sigs/image-builder/pull/1175 because they were not well maintained (I was one of the guilty parties!)

**Comment by lentzi90** [comment]
The exception request has been rejected: https://github.com/cncf/foundation/issues/625#issuecomment-2127628513

**Comment by AverageMarcus** [comment]
Yeah. Quite disappointing. 😞 

I'm talking with the other maintainers to decide how we proceed. Once we have something we'll update.

**Comment by AverageMarcus** [comment]
Not including Packer in the container image would mean the container isn't actually usable as Packer is required for image-builder to function. We're considering this but it would impact our end users that rely on this image to make use of image-builder so we're reluctant to do so until needed.

Once way I think we might be able to get around this (though I'm not 100% sure if this avoids the license or not) is if we have the dependencies fetched at runtime when launching the container image. This could possibly have implications for users that run it in more locked down environments and could cause the image to stop working if an external dependency isn't available at runtime.

**Comment by BenTheElder** [comment]
I would avoid any sort of runtime dependency that does not meet CNCF licensing requirements, regardless of distribution or linkage.


So:
- tool we ship to user fetches at runtime ❌ 
- tool we run in CI *for development purposes* fetches at runtime ✅ 

(e.g. we fetch shellcheck in kubernetes CI, but users of the Kubernetes release are not expected to use this, and we do NOT host it)

but the ultimate authority on this is CNCF / LF legal.

> There has been some [rumours that the IBM purchase of HashiCorp could possibly lead to them reverting the license change](https://www.theregister.com/2024/05/24/opinion_column_terraform/) back to a real open source license. The likelihood of this is admittedly very low I think but it would be ideal if that is the outcome so we're hoping we can "buy some time" until that is either confirmed or not.

We're not depending on this elsewhere in Kubernetes and I wouldn't bet on it, given no intention has been announced. It also may be quite some time before the deal closes even if there were clear intent.

Forking pre-BUSL packer could be acceptable but sounds difficult to sustain and probably should be discussed as part of a larger effort under the CNCF or similar.

**Comment by AverageMarcus** [comment]
Maybe I'm misunderstanding @BenTheElder but it sounds like you're suggesting that we should stop our reliance on Packer completely, is that correct?

**Comment by BenTheElder** [comment]
I don't think that's the only valid option but I also don't think
> Once way I think we might be able to get around this (though I'm not 100% sure if this avoids the license or not) is if we have the dependencies fetched at runtime when launching the container image. 

Is a workable approach.

The CNCF recommends freezing at the current version as a short term approach but I think either an exception or a fork will be necessary for long term:
https://github.com/cncf/foundation/blob/main/source-available-recommendations.md#recommendations

If there aren't sufficient resources for a fork (or switching tools), you could request an exception (see doc above) and maybe also seek clarification if the "we download it at runtime" approach is acceptable.

**Comment by AverageMarcus** [comment]
> If there aren't sufficient resources for a fork (or switching tools), you could request an exception

We already tried that and it was rejected. 😔 See https://github.com/cncf/foundation/issues/625#issuecomment-2127628513

**Comment by fabriziopandini** [comment]
I tried to give my 2 cents [here](https://github.com/cncf/foundation/issues/625#issuecomment-2137588297)

TL;DR;
I think the usage the project is doing do of packer should not be classified a runtime dependency, because it can lead to confusion.

AFAIK, the project is using packer as a build tool, and packer is not included at all in any of the images (OVA, ami etc) which are the ultimate output of the project

The "commodity docker image" which is publiished as a release artifact, instead is just a commodity to avoid to install dependency locally, not the core output of the project; I understand that this is not ideal, but in case this can help in addressing the issue, IMO the image builder project can consider stopping publishing those "commodity images". Users can always install the required tools locally, or eventually build the "commodity image" locally if they prefer.

**Comment by mnaser** [comment]
I'd like to share a very simple approach that has worked really well for us and is mostly "already a solution":

https://github.com/vexxhost/capo-image-elements

With diskimage-builder, a project that has build thousands of images and the concept of elements, it quite simple to build an image ready for CAPI.  The only caveat is that it doesn't build against a cloud, but it builds locally instead in a chroot, and then gives you a raw/qcow2 file.  You would then need to take that artifact and push that into wherever.

We have many customers using it, we use it in our product, and it already works really well.  It is also likely very easy to customize.

**Comment by sathieu** [comment]
Couldn't you build on top of [OpenStack's Diskimage-builder](https://docs.openstack.org/diskimage-builder/latest/index.html)?

EDIT: it's already used by metal3, see [here](https://github.com/metal3-io/project-infra/tree/main/jenkins/image_building)

**Comment by AverageMarcus** [comment]
@sathieu Does that support Windows images? Does it support ignition-based OS's? Can it be used as a drop-in replacement for Packer? (If not someone is going to need to rebuild everything we've currently got and none of the maintainers have that kind of time available unfortunately 😞 )

**Comment by sathieu** [comment]
> Does that support Windows images?

No.

> Does it support ignition-based OS's?

There is no Flatcar. But this could probably be done, because fedora and rhel are supported.

> Can it be used as a drop-in replacement for Packer? (If not someone is going to need to rebuild everything we've currently got and none of the maintainers have that kind of time available unfortunately 😞 )

No, but I can help.

**Comment by mboersma** [comment]
See #2025 for a way to use a newer Packer version in your pipelines (just not here in a CNCF project).
